### PR TITLE
Mock Fusion instance correctly in IndexStageTestBase

### PR DESCRIPTION
**Pull request by krook1024**
# Tl;dr

Mock Fusion instance correctly in IndexStageTestBase

# Details

Previously, if the `IndexStage` you were testing contained a call to `IndexStage::newDocument`, the unit test would fail because of an NPE because the tested `IndexStage` environment tries to create a document, but it cannot because the behaviour of the `Fusion` instance that's injected into the `IndexStage` isn't mocked correctly.

This PR fixes that behaviour and lets IndexStage instances run newDocuments, but it runs the IndexStageTestBase::newDocument in the background.

# How to verify

Let's suppose you have an IndexStage that looks something like this:

```java
public class MyStage extends IndexStageBase<MyStageConfig> {
// [...]
    @Override
    public void process(Document document, Context context, Consumer<Document> output) {
       Document document = newDocument();
       output.accept(document);
    }
// [...]
}

```

And the test looking like this:

```java
public class MyStageTest extends IndexStageTestBase<MyStageConfig> {
// [...]
 @Test
    public void testWhenValidUrlIsProvidedDocumentShouldBeFed() {
         MyStageConfig conf = newConfig(MyStageConfig.class, config -> {
             // [...]
        }); 

        MyStage underTest = createStage(MyStage.class, conf);

        underTest.process(doc, null, System.out::println);
    }
// [...]
}
```

It would end up in a NPE in the current version, but with this PR it should print the new document.

### CODE REVIEW

**Comment by krook1024** [comment]
@dzmitryk @sbrowndev can you guys take a look this?
